### PR TITLE
[Collage] SubGraphs

### PR DESCRIPTION
**Pull request by mbs-octoml**
See https://github.com/apache/tvm-rfcs/blob/main/rfcs/0062-collage.md.

Collage works in units of 'sub-graphs', which are potential partitions of the
overall Relay model. This PR introduces SubGraph (an arbitrary partitioning, without
any implication about how it is to be represented), it's companion SubSubGraph
(implying a representation as a function), and some supporting odds 'n ends.

### CODE REVIEW

**Comment by mbs-octoml** [comment]
@mbaret, thanks.

**Comment by mbs-octoml** [comment]
green and ready for review

File: src/relay/collage/dataflow_graph.h [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    ICHECK_LT(index, indexed_graph_->size());
+    return downstream_map_[index];
+  }
+
+ private:
+  /*! \brief The overall expression. */
+  Expr expr_;
+  /*! \brief The indexed graph which captures the main dataflow. */
+  std::unique_ptr<IndexedGraph<Expr>> indexed_graph_;
+  /*! \brief Map from a node's PostDfsIndex to the set of it's downstream dataflow node indexes. */
```
Comment (SebastianBoblest) [inline src/relay/collage/dataflow_graph.h:69]:
```suggestion
  /*! \brief Map from a node's PostDfsIndex to the set of its downstream dataflow node indexes. */
```

File: src/relay/collage/sub_graph.cc [hunk 2]
```diff
@@ -0,0 +1,1032 @@
...
+      result.Add(matcher.expr_to_node(matched_sub_expr)->index_);
+    }
+  }
+  return result;
+}
+
+std::string SubGraphConfig::ToString() const {
+  std::ostringstream os;
+  os << "{max_exits=" << max_exits;
+  os << ",allow_taps=" << allow_taps;
```
Comment (SebastianBoblest) [inline src/relay/collage/sub_graph.cc:480]:
```suggestion
  os << ", allow_taps=" << allow_taps;
```

File: src/relay/collage/sub_graph.cc [hunk 3]
```diff
@@ -0,0 +1,1032 @@
...
+  os << ",max_max_depth=" << max_max_depth;
```
Comment (SebastianBoblest) [inline src/relay/collage/sub_graph.cc:481]:
```suggestion
  os << ", max_max_depth=" << max_max_depth;
```

File: src/relay/collage/sub_graph.cc [hunk 4]
```diff
@@ -0,0 +1,1032 @@
...
+size_t SubSubGraphNode::hash() const {
+  size_t h = StructuralHash()(attrs_);
+  h ^= sub_graph()->hash() + 0x9e3779b9 + (h << 6) + (h >> 2);
+  return h;
+}
+
+std::string SubSubGraphNode::ToString() const {
+  std::ostringstream os;
+  os << "{sub_graph=" << sub_graph()->ToString();
+  os << ",attrs=" << PrettyPrint(attrs_);
```
Comment (SebastianBoblest) [inline src/relay/collage/sub_graph.cc:511]:
```suggestion
  os << ", attrs=" << PrettyPrint(attrs_);
```

Comment (SebastianBoblest) [reply src/relay/collage/sub_graph.cc:511]: It seems you intentionally omit the whitespace, so please ignore comment.

File: src/relay/collage/sub_graph.cc [hunk 5]
```diff
@@ -0,0 +1,1032 @@
...
+        // We found a deeper path to an already expanded node. We'll expand again.
+        max_depths[output_node] = next_depth;
+        stack.push_back(output_node);
+      }
+    }
+  }
+  return max_depth;
+}
+
+/*! \brief Return's true if any (input/output) of node is (outside/inside) the sub-graph.  */
```
Comment (SebastianBoblest) [inline src/relay/collage/sub_graph.cc:797]:
```suggestion
/*! \brief Returns true if any (input/output) of node is (outside/inside) the sub-graph.  */
```

File: src/relay/collage/sub_graph.cc [hunk 6]
```diff
@@ -0,0 +1,1032 @@
...
+  Expr result = dataflow_graph.expr();
+  for (const auto& sub_graph : sub_graphs) {
+    result = sub_graph->Rewrite(dataflow_graph, result);
+  }
+  return result;
+}
+
+/*!
+ * \brief A pass which partitions (the unique) global function in the module according to the
+ * post-dfs indexes in \p indexes. The partiting must respect the configuration with \p max_exits
```
Comment (SebastianBoblest) [inline src/relay/collage/sub_graph.cc:946]:
```suggestion
 * post-dfs indexes in \p indexes. The partitioning must respect the configuration with \p max_exits
``

File: src/relay/collage/sub_graph.h [hunk 7]
```diff
@@ -0,0 +1,451 @@
...
+
+namespace tvm {
+namespace relay {
+namespace collage {
+
+/*! \brief Returns operator pattern kind as single-letter string. */
+std::string KindToString(OpPatternKind kind);
+
+/*!
+ * \brief Returns a kind and label for the single \p sub_expr, ignoring it's sub-sub expressions.
```
Comment (SebastianBoblest) [inline src/relay/collage/sub_graph.h:49]:
```suggestion
 * \brief Returns a kind and label for the single \p sub_expr, ignoring its sub-sub expressions.
```

File: src/relay/collage/sub_graph.h [hunk 8]
```diff
@@ -0,0 +1,451 @@
...
+  IndexSet Downstream(const DataflowGraph& dataflow_graph) const;
+
+  /*!
+   * \brief Returns true if this sub-graph is valid. Ie:
+   *  - no output of the sub-graph can flow to any input of the sub-graph (otherwise we'd end up
+   *    with a dataflow cycle when we partition).
+   *  - all inputs and outputs of the sub-graph are in the same scope, ie not separated by
+   *    control flow (otherwise there'd be no consistent program point at which to eval the
+   *    partitioned function).
+   *  - no more than config.max_outputs outputs are require.
```
Comment (SebastianBoblest) [inline src/relay/collage/sub_graph.h:324]:
```suggestion
   *  - no more than config.max_outputs outputs are required.
```

File: src/relay/collage/sub_graph.h [hunk 9]
```diff
@@ -0,0 +1,451 @@
...
+ *
+ * Extraction yields a function with input nodes replaced by parameters and exit nodes in the
+ * function result. Rewriting replaces the sub-graph with a call to that function, and all
+ * outputs with (projections from) the call result.
+ *
+ * (Note that it's tempting to move attrs_ into \p SubGraphNode and thus avoid this class.
+ * However we found the implementation was easier to understand in this form since it makes
+ * the result of \p Extract unambiguous.)
+ */
+class SubSubGraphNode : public Object {
```
Comment (mbaret) [inline src/relay/collage/sub_graph.h:99]: I think 'NestedSubGraph' is a clearer name for this class.

**Review by mbaret — COMMENTED** [review]
Largely looks good to me, couple of points around naming. It's tacitly acknowledged that more of this infrastructure will get exercised by tests in later PRs.

File: src/relay/collage/sub_graph.cc (hunk 4 above)
Comment (mbs-octoml) [reply src/relay/collage/sub_graph.cc:511]: The debug output is pretty overwhelming so I was trying to compress it a bit but I went ahead and took your suggestion throughout anyway :-)

File: src/relay/collage/README.md [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+
+The `CollagePartition` pass for finding optimal partitionings of Relay models.
+
+See the [RFC](https://github.com/mbs-octoml/mbs-tvm-rfcs/blob/mbs-rfcs-collage/rfcs/xxxx-collage.md).
+
+Based on:
+> *Collage: Automated Integration of Deep Learning Backends*  
+> Byungsoo Jeon, Sunghyun Park, Peiyuan Liao, Sheng Xu, Tianqi Chen, Zhihao Jia
+
+CAUTION: This is a prototype, do not use in prod.
```
[the remark this answers is not in the document]
Comment (mbs-octoml) [reply src/relay/collage/README.md:26]: Decided to leave it in there given the unit testing is pretty thin on the ground and we've only been exercising on a handful of models.

File: src/relay/collage/sub_graph.h [hunk 11]
```diff
@@ -0,0 +1,451 @@
...
+   * inside this sub-sub-graph must correspond to nodes shared between \p dataflow_graph.expr() and
+   * \p expr. See \p SubGraph::ParallelRewrite below.
+   */
+  Expr Rewrite(const DataflowGraph& dataflow_graph, const Expr& expr) const;
+
+  static constexpr const char* _type_key = "relay.collage.SubSubGraph";
+  TVM_DECLARE_FINAL_OBJECT_INFO(SubSubGraphNode, Object);
+};
+
+class SubSubGraph : public ObjectRef {
```
[the remark this answers is not in the document]
Comment (mbs-octoml) [reply src/relay/collage/sub_graph.h:140]: I hadn't thought of that actually. At first I had SubGraph be directly recursive until I realized things were much clearer with the intermediate NestedSubGraph, and was so happy with that I didn't push further. You are right there's some signature sharing but no implementation sharing I can see, and I think making code polymorphic on SubGraph vs NestedSubGraph would only make things even more confusing. So let me leave it as is.

File: src/relay/collage/sub_graph.h [hunk 12]
```diff
@@ -0,0 +1,451 @@
...
+  bool operator!=(const SubGraphNode& that) const { return !(*this == that); }
+  bool operator<(const SubGraphNode& that) const;
+  size_t hash() const;
+
+ private:
+  /*! \brief Initialize the entry/exit/input/output sets given the inside and \p dataflow_graph. */
+  void Init(const DataflowGraph& dataflow_graph);
+
+  /*! \brief Calculates and returns the maximum path depth. */
+  size_t MaxDepth(const DataflowGraph& dataflow_graph) const;
```
[the remark this answers is not in the document]
Comment (mbs-octoml) [reply src/relay/collage/sub_graph.h:357]: From now on every time I come up with some awkward stuttering variable name I'll thing "what would Matthew write?" :-)

**Review by mbaret — APPROVED** [review]
LGTM

**Pull request merged**